src/metabase/util/time/impl.cljs

```diff
@@ -262,17 +296,48 @@
                (or date? date-time?) (coerce-local-date-time input))]
        (if (and t (.isValid t))
          (or
-          (format-extraction-unit t unit)
+          (format-extraction-unit t unit locale)
+          ;; no locale for default formats
           (cond
             time? (.format t "h:mm A")
             date? (.format t "MMM D, YYYY")
             date-time? (.format t "MMM D, YYYY, h:mm A")))
          input))
+
+     (number? input)
      (if (= unit :hour-of-day)
        (str (cond (zero? input) "12" (<= input 12) input :else (- input 12)) " " (if (<= input 11) "AM" "PM"))
        (or
-        (format-extraction-unit (common/number->timestamp input {:unit unit}) unit)
-        (str input))))))
+        (format-extraction-unit (common/number->timestamp input {:unit unit}) unit locale)
+        (str input)))
+
+     (moment/isMoment input)
+     (or (format-extraction-unit input unit locale)
+         ;; no locale for default formats
+         (cond
+           ;; no hour, minute, or seconds, must be date
+           (not (has-explicit-time? input))
+           (.format input "MMM D, YYYY")
+
+           ;; no year, month, or day, must be a time
+           (not (has-explicit-date? input))
+           (.format input "h:mm A")
+
+           :else ;; otherwise both date and time
+           (.format input "MMM D, YYYY, h:mm A"))))))
+
+(defn parse-unit
+  "Parse a unit of time/date, e.g., 'Wed' or 'August' or '14'."
+  ([input unit]
+   (when-some [format (get unit-formats unit)]
+     (moment input format)))
+  ([input unit locale]
+   (let [temp (.locale moment)]     ;; 1. save current locale
+     (try
+       (.locale moment locale)      ;; 2. set new locale for subsequent parse
+       (parse-unit input unit)
+       (finally
+         (.locale moment temp)))))) ;; 3. set locale to original
```

Comment:
`temp` is misleading here. It's actually the opposite - we're saving the `original` locale, setting it to `locale` temporarily, and then putting the `original` back.  I think `saved` or `original` is the right name.